src/ListExecutionManager.php

```diff
@@ -0,0 +1,132 @@
+<?php
+
+declare(strict_types = 1);
+
+namespace Drupal\oe_list_pages;
+
+use Drupal\Core\Entity\EntityRepositoryInterface;
+use Drupal\Core\Entity\EntityTypeManager;
+use Drupal\Core\Language\LanguageManagerInterface;
+use Drupal\Core\Pager\PagerManagerInterface;
+use Symfony\Component\HttpFoundation\RequestStack;
+
+/**
+ * Handles the list pages execution.
+ */
+class ListExecutionManager implements ListExecutionManagerInterface {
+
+  /**
+   * The list source factory.
+   *
+   * @var \Drupal\oe_list_pages\ListSourceFactoryInterface
+   */
+  protected $listSourceFactory;
+
+  /**
+   * The request stack.
+   *
+   * @var \Symfony\Component\HttpFoundation\RequestStack
+   */
+  protected $requestStack;
+
+  /**
+   * The entity type manager.
+   *
+   * @var \Drupal\Core\Entity\EntityTypeManager
+   */
+  protected $entityTypeManager;
+
+  /**
+   * The pager manager.
+   *
+   * @var \Drupal\Core\Pager\PagerManagerInterface
+   */
+  protected $pagerManager;
+
+  /**
+   * The pager.
+   *
+   * @var \Drupal\Core\Pager\PagerManagerInterface
+   */
+  protected $pager;
+
+  /**
+   * The entity repository.
+   *
+   * @var \Drupal\Core\Entity\EntityRepositoryInterface
+   */
+  protected $entityRepository;
+
+  /**
+   * The language manager.
+   *
+   * @var \Drupal\Core\Language\LanguageManagerInterface
+   */
+  protected $languageManager;
+
+  /**
+   * ListManager constructor.
+   *
+   * @param \Drupal\oe_list_pages\ListSourceFactoryInterface $listSourceFactory
+   *   The list source factory.
+   * @param \Drupal\Core\Entity\EntityTypeManager $entityTypeManager
+   *   The entity type manager.
+   * @param \Drupal\Core\Pager\PagerManagerInterface $pager
+   *   The pager.
+   * @param \Drupal\Core\Entity\EntityRepositoryInterface $entityRepository
+   *   The entity repository.
+   * @param \Symfony\Component\HttpFoundation\RequestStack $requestStack
+   *   The request stack.
+   * @param \Drupal\Core\Language\LanguageManagerInterface $languageManager
+   *   The language manager.
+   */
+  public function __construct(ListSourceFactoryInterface $listSourceFactory, EntityTypeManager $entityTypeManager, PagerManagerInterface $pager, EntityRepositoryInterface $entityRepository, RequestStack $requestStack, LanguageManagerInterface $languageManager) {
+    $this->listSourceFactory = $listSourceFactory;
+    $this->entityTypeManager = $entityTypeManager;
+    $this->pager = $pager;
+    $this->entityRepository = $entityRepository;
+    $this->requestStack = $requestStack;
+    $this->languageManager = $languageManager;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function executeList($entity): ?ListExecutionResults {
+
+    static $executed_lists = [];
```

Comment:
It's called by different components (list results, form, will be probably also on pager and similars) thouhgh the need to prevent repeated executions.